hibernate-core/src/main/java/org/hibernate/generator/Generator.java

```diff
@@ -93,6 +96,32 @@ public interface Generator extends Serializable {
 	 */
 	boolean generatedOnExecution();
 
+
+	/**
+	 * Determines if the property value is generated when a row is written to the database,
+	 * or in Java code that executes before the row is written.
+	 * <p>
+	 * Defaults to {@link #generatedOnExecution()}, but can be overloaded allowing conditional
+	 * value generation timing (on/before execution) based on the current state of the owner entity.
+	 * Note that a generator <b>must</b> implement both {@link BeforeExecutionGenerator} and
+	 * {@link OnExecutionGenerator} to achieve this behavior.
+	 *
+	 * @param session The session from which the request originates.
+	 * @param entity The instance of the entity owning the attribute for which we are generating a value.
+	 *
+	 * @return {@code true} if the value is generated by the database as a side effect of
+	 * the execution of an {@code insert} or {@code update} statement, or false if
+	 * it is generated in Java code before the statement is executed via JDBC.
+	 *
+	 * @see #generatedOnExecution()
+	 * @see BeforeExecutionGenerator
+	 * @see OnExecutionGenerator
+	 * @since 6.4
+	 */
+	default boolean generatedOnExecution(SharedSessionContractImplementor session, Object entity) {
```

Comment:
Usually, we pass context objects like `session` last, but that's a minor note.

Reply:
Thanks for the suggestion, inverted the method parameters.